prusti-specs/src/type_model/mod.rs

```diff
@@ -0,0 +1,392 @@
+//! Processes `#[model]` attributed types.
+//!
+//! Usage documentation can be found in the corresponding macro definition.
+//!
+//! Given a `#[model]` attributed type `T`, this logic creates the following three items:
+//! * A struct `M` which holds the model's fields
+//! * A trait which provides a `model` method to be used in specifications
+//! * An implementation of the aforementioned trait for `T`.
+//!   The implementation is `unimplemented!()`, `#[pure]` and `#[trusted]`
+//!
+//! The model struct `M` must be copyable.
+//!
+//! # Note
+//! This macro always generates a trait with a `model` method on the fly for every modelled type.
+//! With this design, one can even model external types which are not present in the local crate.
+
+use super::parse_quote_spanned;
+use proc_macro2::{Ident, TokenStream};
+use quote::ToTokens;
+use syn::{parse_quote, spanned::Spanned};
+
+/// See module level documentation
+pub fn rewrite(item_struct: syn::ItemStruct) -> syn::Result<TokenStream> {
+    let res = rewrite_internal(item_struct);
+    match res {
+        Ok(result) => Ok(result.into_token_stream()),
+        Err(err) => Err(err.into()),
+    }
+}
+
+type TypeModelGenerationResult<R> = Result<R, TypeModelGenerationError>;
+
+fn rewrite_internal(item_struct: syn::ItemStruct) -> TypeModelGenerationResult<TypeModel> {
+    let idents = GeneratedIdents::generate(&item_struct);
+    let model_struct = create_model_struct(&item_struct, &idents)?;
+    let to_model_trait = create_to_model_trait(&item_struct, &model_struct, &idents);
+    let model_impl = create_model_impl(&item_struct, &model_struct, &to_model_trait)?;
+
+    Ok(TypeModel {
+        model_struct,
+        model_impl,
+        to_model_trait,
+    })
+}
+
+fn create_model_struct(
+    item_struct: &syn::ItemStruct,
+    idents: &GeneratedIdents,
+) -> TypeModelGenerationResult<syn::ItemStruct> {
+    if item_struct.fields.is_empty() {
+        return Err(TypeModelGenerationError::MissingStructFields(
+            item_struct.span(),
+        ));
+    }
+
+    let model_struct_ident = &idents.model_struct_ident;
+    let mut model_struct: syn::ItemStruct = parse_quote_spanned! {item_struct.span()=>
+        #[derive(Copy, Clone)]
```

Comment:
`PartialEq` and `Eq` as well?

Reply:
Simply deriving those two traits will have no immediate effect since we cannot assume a specification for `PartialEq::eq`.
We could generate the specification during the macro expansion, but it might be overkill?

Reply:
IIRC we treat types with derived `PartialEq` and `Eq` specially, because we use [`has_structural_eq_impls`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.has_structural_eq_impls)?